File: puzzles/puzzle_02_come_fly_with_me.py
"""
"Come Fly With Me"
by Not Quite Tangible

https://www.reddit.com/r/BloodOnTheClocktower/comments/1ewxu0r/weekly_puzzle_2_come_fly_with_me/
"""

from __future__ import annotations

from botc_solver import BOTCModel, CharacterType, print_solution
from botc_solver.characters import (
    Balloonist,
    Clockmaker,
    Drunk,
    FortuneTeller,
    FortuneTellerCheck,
    Goblin,
    Investigator,
    Juggler,
    Knight,
    Leviathan,
    Saint,
    Seamstress,
    apply_claims,
    player_names,
    script,
)
from botc_solver.core import role_character_type

PLAYERS = [
    Investigator(name="Sarah", role=Goblin, among=["Matthew", "Fraser"]),
    Juggler(
        name="Matthew",
        guesses={
            "Steph": Knight,
            "Sarah": Leviathan,
            "Anna": Goblin,
            "Sula": Goblin,
            "You": Seamstress,
        },
        correct_count=2,
    ),
    Clockmaker(name="Anna", demon_next_to_minion=True),
    Balloonist(
        name="Sula",
        different_character_type_pairs=[
            ("Tim", "Matthew"),
            ("Matthew", "Steph"),
        ],
    ),
    Seamstress(name="You", aligned=True, among=["Matthew", "Sula"]),
    Knight(name="Steph", no_demon_among=["Tim", "Sula"]),
    FortuneTeller(
        name="Fraser",
        checks=[
            FortuneTellerCheck(left="Sarah", right="Anna", yes=False),
            FortuneTellerCheck(left="You", right="Fraser", yes=False),
            FortuneTellerCheck(left="Steph", right="Sarah", yes=False),
        ],
    ),
    Saint(name="Tim"),
]
PLAYER_NAMES = player_names(PLAYERS)
CHARACTERS = script(
    Leviathan,
    Goblin,
    Drunk,
    Saint,
    Balloonist,
    Clockmaker,
    FortuneTeller,
    Investigator,
    Juggler,
    Knight,
    Seamstress,
)


def build_model() -> BOTCModel:
    game = BOTCModel(PLAYER_NAMES, characters=CHARACTERS, seating=PLAYER_NAMES)

    game.set_character_count(Leviathan, 1)
    game.set_character_count(Goblin, 1)
    game.fix_not_actual("You", Leviathan)
    game.fix_not_actual("You", Goblin)

    for player in PLAYER_NAMES:
        game.fix_poisoned(player, False)

    outsider_count = sum(
        game.actual_is(player, role)
        for player in PLAYER_NAMES
        for role, character in game.characters.items()
        if role_character_type(character) == CharacterType.OUTSIDER
    )
    balloonist_in_play = game.role_in_play(Balloonist)
    game.add_enforced(outsider_count == 1, balloonist_in_play.Not())
    game.add_enforced(outsider_count >= 1, balloonist_in_play)
    game.add_enforced(outsider_count <= 2, balloonist_in_play)

    apply_claims(game, PLAYERS)

    return game


def solve():
    return build_model().solve_all()


def main() -> None:
    print_solution(
        solve(),
        PLAYER_NAMES,
        forced_roles=[Goblin, Leviathan, Drunk],
        forced_missing="not in every world",
    )


if __name__ == "__main__":
    main()

File: typescript/src/puzzles/puzzle-02-come-fly-with-me.ts
import { BOTCModel } from "../model";
import { printSolution } from "../display";
import { CharacterType } from "../core";
import {
  Balloonist,
  Clockmaker,
  Drunk,
  FortuneTeller,
  Goblin,
  Investigator,
  Juggler,
  Knight,
  Leviathan,
  Saint,
  Seamstress,
  applyClaims,
  playerNames,
  script,
} from "../characters";
import { roleCharacterType } from "../core";

export const PLAYERS = [
  new Investigator({ name: "Sarah", role: Goblin, among: ["Matthew", "Fraser"] }),
  new Juggler({
    name: "Matthew",
    guesses: { Steph: Knight, Sarah: Leviathan, Anna: Goblin, Sula: Goblin, You: Seamstress },
    correctCount: 2,
  }),
  new Clockmaker({ name: "Anna", demonNextToMinion: true }),
  new Balloonist({
    name: "Sula",
    differentCharacterTypePairs: [
      ["Tim", "Matthew"],
      ["Matthew", "Steph"],
    ],
  }),
  new Seamstress({ name: "You", aligned: true, among: ["Matthew", "Sula"] }),
  new Knight({ name: "Steph", noDemonAmong: ["Tim", "Sula"] }),
  new FortuneTeller({
    name: "Fraser",
    checks: [
      { left: "Sarah", right: "Anna", yes: false },
      { left: "You", right: "Fraser", yes: false },
      { left: "Steph", right: "Sarah", yes: false },
    ],
  }),
  new Saint({ name: "Tim" }),
];

export const PLAYER_NAMES = playerNames(PLAYERS);
export const CHARACTERS = script(
  Leviathan,
  Goblin,
  Drunk,
  Saint,
  Balloonist,
  Clockmaker,
  FortuneTeller,
  Investigator,
  Juggler,
  Knight,
  Seamstress,
);

export function buildModel(): BOTCModel {
  const game = new BOTCModel(PLAYER_NAMES, { characters: CHARACTERS, seating: PLAYER_NAMES });
  game.setCharacterCount(Leviathan, 1);
  game.setCharacterCount(Goblin, 1);
  game.fixNotActual("You", Leviathan);
  game.fixNotActual("You", Goblin);
  const outsiderVars = PLAYER_NAMES.flatMap((player) =>
    [...game.characters.entries()]
      .filter(([, character]) => roleCharacterType(character) === CharacterType.Outsider)
      .map(([role]) => game.actualIs(player, role)),
  );
  const balloonistInPlay = game.roleInPlay(Balloonist);
  game.addEnforcedExactlyN(outsiderVars, 1, balloonistInPlay.not());
  game.addEnforcedAtLeastN(outsiderVars, 1, balloonistInPlay);
  game.addEnforcedAtMostN(outsiderVars, 2, balloonistInPlay);
  applyClaims(game, PLAYERS);
  return game;
}

export async function solve() {
  return buildModel().solveAll();
}

if (import.meta.main && process.argv[1]?.endsWith("puzzle-02-come-fly-with-me.ts"))
  printSolution(await solve(), PLAYER_NAMES, {
    forcedRoles: [Goblin, Leviathan, Drunk],
    forcedMissing: "not in every world",
  });
